src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java

```diff
@@ -0,0 +1,141 @@
+            return ImmutableSet.of(LootContextParams.BLOCK_ENTITY);
+        }
+    };
+    public static final ImprovedContextNbtProvider BLOCK_ENTITY = new ImprovedContextNbtProvider(BLOCK_ENTITY_PROVIDER);
+    final ImprovedContextNbtProvider.Getter getter;
+
+    private static ImprovedContextNbtProvider.Getter forEntity(final LootContext.EntityTarget pEntityTarget) {
+        return new ImprovedContextNbtProvider.Getter() {
+            @Nullable
+            public Tag get(LootContext p_165589_) {
+                Entity entity = p_165589_.getParamOrNull(pEntityTarget.getParam());
+                return entity != null ? NbtPredicate.getEntityTagToCompare(entity) : null;
+            }
+
+            public String getId() {
+                return pEntityTarget.getName();
+            }
+
+            public Set<LootContextParam<?>> getReferencedContextParams() {
+                return ImmutableSet.of(pEntityTarget.getParam());
+            }
+        };
+    }
+
+    private ImprovedContextNbtProvider(ImprovedContextNbtProvider.Getter pGetter) {
+        this.getter = pGetter;
+    }
+
+    public LootNbtProviderType getType() {
+        return EIOLootNbtProvider.IMPROVED_CONTEXT.get();
+    }
+
+    @Nullable
+    public Tag get(LootContext pLootContext) {
+        return this.getter.get(pLootContext);
+    }
+
+    public Set<LootContextParam<?>> getReferencedContextParams() {
+        return this.getter.getReferencedContextParams();
+    }
+
+    public static NbtProvider forContextEntity(LootContext.EntityTarget pEntityTarget) {
+        return new ImprovedContextNbtProvider(forEntity(pEntityTarget));
+    }
+
+    static ImprovedContextNbtProvider createFromContext(String pTargetName) {
+        if (pTargetName.equals(BLOCK_ENTITY_ID)) {
+            return new ImprovedContextNbtProvider(BLOCK_ENTITY_PROVIDER);
+        } else {
+            LootContext.EntityTarget lootcontext$entitytarget = LootContext.EntityTarget.getByName(pTargetName);
+            return new ImprovedContextNbtProvider(forEntity(lootcontext$entitytarget));
+        }
+    }
+
+    interface Getter {
+        @Nullable
+        Tag get(LootContext pLootContext);
+
+        String getId();
+
+        Set<LootContextParam<?>> getReferencedContextParams();
+    }
+
+    public static class InlineSerializer implements GsonAdapterFactory.InlineSerializer<ImprovedContextNbtProvider> {
+        public JsonElement serialize(ImprovedContextNbtProvider p_165597_, JsonSerializationContext p_165598_) {
+            return new JsonPrimitive(p_165597_.getter.getId());
+        }
+
+        public ImprovedContextNbtProvider deserialize(JsonElement p_165603_, JsonDeserializationContext p_165604_) {
+            String s = p_165603_.getAsString();
+            return ImprovedContextNbtProvider.createFromContext(s);
+        }
+    }
+
+    public static class Serializer
+            implements net.minecraft.world.level.storage.loot.Serializer<ImprovedContextNbtProvider> {
```

Comment:
put on one line, rename Serializer to JsonSeriailzer or something, so you can import it